designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,950 @@
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
+    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
+1. When a glob pattern is passed directly (such as `eslint src/*.js`):
+    1. ESLint expands the glob pattern to get a list of files.
+    1. Each file is checked individually as in step 1.
+1. When a directory is passed directly (such as `eslint src`):
+    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
+    1. The glob pattern is checked as in step 2.
+1. When a relative directory is passed directly (such as `eslint .`):
+    1. The relative directory pattern is resolved to a full directory name.
+    1. The glob pattern is checked as in step 3.
+
+**Note:** ESLint will continue to ignore `node_modules` by default.
+
+### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`
+
+Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.
+
+### Implementation Details
+
+The implementation of this feature requires the following changes:
+
+1. Create a new `ConfigArray` class to manage configs.
+1. Create a `--no-config-file` CLI option and alias it to `--no-eslintrc` for backwards compatibility.
+1. Create a `useConfigFile` option for `CLIEngine`. Alias `useEslintrc` to this option for backwards compatibility.
+1. In `CLIEngine#executeOnFiles()`:
+    1. Check for existence of `eslint.config.js`, and if found, opt-in to new behavior.
+    1. Create a `ConfigArray` to hold the configuration information and to determine which files to lint (in conjunction with already-existing `globUtils`)
+    1. Rename the private functions `processText()` and `processFiles()` to `legacyProcessText()` and `legacyProcessFiles()`; create new versions with the new functionality named `processText()` and `processFiles()`. Use the appropriate functions based on whether or not the user has opted-in.
+    1. Update `Linter#verify()` to check for objects on keys that now support objects instead of strings (like `parser`).
```

Comment:
I think we need a new option for `Linter#verify()` in order to disable the processing of `env`. All of the `env` processing is in `Linter#verify()`.

Reply:
Good catch! I'll get that incorporated.